nomad/job_endpoint_hooks.go

```diff
@@ -209,11 +226,118 @@ func (jobImpliedConstraints) Mutate(j *structs.Job) (*structs.Job, []error, erro
 		if ok := consulServiceDisco[tg.Name]; ok {
 			mutateConstraint(constraintMatcherLeft, tg, consulServiceDiscoveryConstraint)
 		}
+
+		// If the task group uses containers, iterate through its tasks
+		if ok := containerPlatform[tg.Name]; ok {
+			// Iterate through all the tasks within the task group and add any required
+			// constraints.
+			for _, t := range tg.Tasks {
+				// If the task uses the docker or podman drivers, figure out its image's
+				// supported architecutres and run the mutator
+				if t.Driver == "docker" || t.Driver == "podman" {
+					refStr, ok := t.Config["image"].(string)
+					if !ok {
+						return nil, nil, ErrNoImage
+					}
+
+					os, arch, err := containerPlatformConstraints(refStr)
+					if err != nil {
+						return nil, nil, err
+					}
+
+					mutateConstraintTask(constraintMatcherLeft, t, os)
+					mutateConstraintTask(constraintMatcherLeft, t, arch)
+				}
+			}
+		}
 	}
 
 	return j, nil, nil
 }
 
+// containerPlatformConstrainsts uses the OCI registry API to detect which platforms
+// are supported by an image or index pointed to by refStr and returns constraints to
+// match those platforms.
+func containerPlatformConstraints(refStr string) (os, arch *structs.Constraint, err error) {
+	ref, err := name.ParseReference(refStr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Get a descriptor for the referenced item
+	desc, err := remote.Head(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if desc.MediaType.IsIndex() {
+		index, err := remote.Index(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		im, err := index.IndexManifest()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Create lists of unique OSes and Architectures
+		oses := map[string]struct{}{}
+		arches := map[string]struct{}{}
+		for _, m := range im.Manifests {
+			oses[m.Platform.OS] = struct{}{}
+			arches[m.Platform.Architecture] = struct{}{}
+		}
+
+		os = &structs.Constraint{
+			LTarget: attrKernelName,
+			Operand: "set_contains_any",
+			RTarget: strings.Join(mapKeys(oses), ","),
+		}
+
+		arch = &structs.Constraint{
+			LTarget: attrCPUArch,
+			Operand: "set_contains_any",
+			RTarget: strings.Join(mapKeys(arches), ","),
+		}
+	} else if desc.MediaType.IsImage() {
+		img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
```

Comment:
It looks to me like we're having the Nomad server authenticate to a remote registry during job submission here? Nomad servers are not typically going to be configured with registry credentials at all, because they don't run workloads themselves.

The approach we're taking here would make it mandatory that all users of Nomad who are using Docker/Podman drivers deploy credentials to their Nomad control plane before upgrading, and have a network topology that allows their control plane access to the registry (which isn't necessarily the case). And it would make all Nomad servers dependent on third-party registries at the time of job submission, which I don't think we can accept as a tradeoff.